feat: encrypt injected env blob for secure serverless deployments (#656)
dmno-dev/varlock: packages/varlock-website/src/content/docs/integrations/vite.mdx

---

## Encrypting the env blob ||encrypting-the-env-blob||

> When using `ssrInjectMode: 'resolved-env'`, varlock injects the fully resolved env data into your SSR build output. By default, this blob is plaintext JSON — meaning anyone with access to the build artifact can read your secrets.

Reviewer: ```suggestion
When using `ssrInjectMode: 'resolved-env'`, Varlock injects the fully resolved env data into your SSR build output. By default, this blob is plaintext JSON — meaning anyone with access to the build artifact can read your secrets.
```

<Steps>

1. **Generate and set the key on your platform**

    For Vercel, this one-liner generates a key and sets it as a sensitive env var for all environments:

    ```bash
    varlock generate-key --plain | vercel env add _VARLOCK_ENV_KEY production preview development --sensitive
    ```

    Or generate and set it manually — the key must be available at both **build time** (for encryption) and **runtime** (for decryption):

    <ExecCommandWidget command="varlock generate-key" showBinary={false} />

2. **Ensure you're using `resolved-env` mode**

    Encryption only applies when the blob is inlined into the build:

    ```ts title="vite.config.ts"
    varlockVitePlugin({ ssrInjectMode: 'resolved-env' })
    ```

3. **Optionally add it to your `.env.local` for local builds**

    ```env title=".env.local"
    _VARLOCK_ENV_KEY=your-64-char-hex-key-here
    ```

</Steps>

:::tip
You can define `_VARLOCK_ENV_KEY` in your `.env.schema` to enable validation. It will automatically be excluded from the injected blob and from type generation.

```env-spec title=".env.schema"
# @sensitive
_VARLOCK_ENV_KEY=
```
:::

:::note

> The encryption key is **never baked into the build**. It must always come from the runtime environment. The encrypted blob is useless without it.

Reviewer: ```suggestion
The encryption key is **never baked into the build**. It must always come from the runtime environment. It is *required* for decryption.
```
